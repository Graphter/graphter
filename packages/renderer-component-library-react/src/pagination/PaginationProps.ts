import { ReactNode } from "react";
import {PaginationPageProps} from "./PaginationPageProps";

export interface PaginationProps {
  page: number
  size: number
  count: number,
  renderPageItem: (props: PaginationPageProps) => ReactNode
}