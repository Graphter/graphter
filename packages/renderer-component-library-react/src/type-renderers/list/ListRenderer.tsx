import React from 'react'
import s from './ListRenderer.pcss'
import { PropertyTypeRendererProps } from "malle-renderer-react";

export default function ListRenderer(
  {
    propertyConfig,
    modelConfig,
    propData,
    modelData,
    setPropDataValue,
    setModelDataValue,
    validationResults
  }: PropertyTypeRendererProps
){
  return (
    <div className={s.listRenderer}>
      List!
    </div>
  )
}