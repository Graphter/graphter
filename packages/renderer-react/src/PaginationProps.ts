export interface PaginationProps {
  skip: number,
  take: number,
  count?: number,
  onChange: (skip: number, take: number) => void
}