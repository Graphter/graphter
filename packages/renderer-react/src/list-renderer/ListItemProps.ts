export interface ListItemProps {
  item?: any,
  index?: number
}