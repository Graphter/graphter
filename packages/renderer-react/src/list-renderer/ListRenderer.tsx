import React, { useEffect, useState, ReactNode, ComponentType, Fragment } from 'react'
import ErrorBoundary from "../error-boundary";
import DefaultError from "../default-error";
import { RenderItemProps } from "./RenderItemProps";
import { RenderPaginationProps } from "./RenderPaginationProps";
import { ErrorRendererProps, NodeConfig } from "@graphter/core";
import serviceStore from "../store/serviceStore";

export interface ListRendererProps {
  config: NodeConfig
  renderItem: ({ item, index}: RenderItemProps) => ReactNode
  renderPagination?: ({ count }: RenderPaginationProps) => ReactNode
  errorRenderer?: ComponentType<{ err: Error | string }>
  page?: number
  size?: number
}

export default function ListRenderer(
  {
    config,
    renderItem,
    renderPagination,
    errorRenderer,
    page,
    size }: ListRendererProps): any {
  const ErrorDisplayComponent: ComponentType<ErrorRendererProps> = errorRenderer || DefaultError;
  if(!config) return <ErrorDisplayComponent err={new Error('Model configuration is required')} />;

  const service = serviceStore.get(config.id);

  const [ count, setCount ] = useState(0);
  const [ items, setItems ] = useState<Array<any>>([]);
  const [ loading, setLoading ] = useState(false);
  const [ error, setError ] = useState<Error>();

  useEffect(() => {
    (async () => {
      setLoading(true);
      let listResult;
      try {
        const take = size || 10;
        const skip = page ? (page - 1) * take : 0;
        listResult = await service.list(skip, take);
      } catch(err){
        setLoading(false);
        setError(err);
        return;
      }
      setItems(listResult.items);
      setCount(listResult.count);
      setLoading(false);
    })()
  }, [ config, page, size ]);

  if(!renderItem) return <ErrorDisplayComponent err={'Missing list item renderer control'} />;

  return (
    <Fragment>
      { error && <ErrorDisplayComponent err={error} /> }
      { loading && <div>loading...</div> }
      <div data-testid='list'>

      {items.map((item, i) => (
        renderItem && (
          <ErrorBoundary
            key={i}
            errorRenderer={ErrorDisplayComponent}>
            {renderWithErrorHandling(() => renderItem({ item }))}
          </ErrorBoundary>)
        ))}
      </div>

      {renderPagination && (
        <ErrorBoundary errorRenderer={ErrorDisplayComponent}>
          {renderWithErrorHandling(() => renderPagination({count}))}
        </ErrorBoundary>
      )}
    </Fragment>
  );

  function renderWithErrorHandling(fn: () => any){
    try {
      return fn();
    } catch(err) {
      return <ErrorDisplayComponent err={err} />
    }
  }
}