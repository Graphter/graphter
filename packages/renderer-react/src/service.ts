const listFns: { [id: string]:  () => Promise<ListResult>} = {};

type ListResult = {
  items: Array<any>,
  count: number,
  skip: number,
  take: number
}

export default {
  init: (modelId: string, fn: () => Promise<ListResult>) => {
    listFns[modelId] = fn;
  },
  list: async (modelId: string): Promise<ListResult> => {
    const fn = listFns[modelId];
    if(!fn) throw new Error(`'${modelId}' doesn't have a list function`);
    return fn();
  }
}
