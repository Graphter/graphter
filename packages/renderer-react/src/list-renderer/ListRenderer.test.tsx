import React from "react";
import service from '../service';
import ListRenderer, {MalleModelConfig} from "./ListRenderer";
import { act, render, within } from '@testing-library/react';
import { screen } from '@testing-library/dom'
import '@testing-library/jest-dom/extend-expect'
import { ListItemProps } from "../ListItemProps";
import {PaginationProps} from "../PaginationProps";

jest.mock('../service');
const mockedService = service as jest.Mocked<typeof service>;

describe('<ListRenderer />', () => {
  let baseModelConfig: MalleModelConfig;
  beforeEach(() => {
    baseModelConfig = {
      id: 'page',
      name: 'Page',
      properties: [
        { id: 'id', type: 'number' },
        { id: 'somethingElse', type: 'string'},
        { id: 'title', type: 'string' },
      ]
    };
  });
  it('should load the first set of items', async () => {
    mockedService.list.mockResolvedValue({ items: [], count: 0, skip: 0, take: 10 });
    await act( async() => {
      render(<ListRenderer
        config={baseModelConfig}
        itemRenderer={() => <div>Item</div>}
        pagination={() => <div>Page</div>} />);
    });
    expect(service.list).toHaveBeenCalledWith('page', 0, 10);
  });
  it('should render items using the supplied item renderer', async () => {
    mockedService.list.mockResolvedValue({
      items: [
        {
          id: 1,
          title: 'some page name',
          somethingElse: 'blah blah blah'
        }
      ],
      count: 0,
      skip: 0,
      take: 10
    });
    const itemRenderer = ({ item }: ListItemProps) => (
      <div key={item.id} data-testid='item-renderer'>
        #{item.id} {item.title} - {item.somethingElse}
      </div>
    )
    await act( async() => {
      render(<ListRenderer
        config={baseModelConfig}
        itemRenderer={itemRenderer}
        pagination={() => <div>Page</div>} />);
    });
    const item = screen.getByTestId('item-renderer');
    expect(item).toMatchSnapshot()
  });
  it('should render items using the supplied pagination control', async () => {
    mockedService.list.mockResolvedValue({
      items: [
        {
          id: 1,
          title: 'some page name',
          somethingElse: 'blah blah blah'
        }
      ],
      count: 0,
      skip: 0,
      take: 10
    });
    const pagination = ({ skip, take, count }: PaginationProps) => (
      <div data-testid='pagination'>
        {skip} - {take} - {count}
      </div>
    )
    await act( async() => {
      render(<ListRenderer
        config={baseModelConfig}
        itemRenderer={() => <div>Item</div>}
        pagination={pagination} />);
    });
    const item = screen.getByTestId('pagination');
    expect(item).toMatchSnapshot()
  });
});

export {};