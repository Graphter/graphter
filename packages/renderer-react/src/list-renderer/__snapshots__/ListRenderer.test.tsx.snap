// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListRenderer /> should render items using the supplied item renderer 1`] = `
<div
  data-testid="item-renderer"
>
  #
  1
   
  some page name
   - 
  blah blah blah
</div>
`;

exports[`<ListRenderer /> should render items using the supplied pagination control 1`] = `
<div
  data-testid="pagination"
>
  0
   - 
  10
   - 
  0
</div>
`;
